client/my-sites/controller.js

```diff
@@ -112,46 +111,62 @@ function renderNoVisibleSites( context ) {
 	);
 }
 
-function isPathAllowedForDomainOnlySite( pathname, domainName ) {
+function renderSelectedSiteIsDomainOnly( { reactContext, selectedSite } ) {
+	const EmptyContentComponent = require( 'components/empty-content' );
+	const { store: reduxStore } = reactContext;
+
+	removeSidebar( reactContext );
+
+	renderWithReduxStore(
+		React.createElement( EmptyContentComponent, {
+			title: i18n.translate( 'This feature is not available for domains' ),
+			line: i18n.translate( 'To use this feature you need to create a site' ),
+			action: i18n.translate( 'Create New Site' ),
+			actionURL: '//dashboard.wordpress.com/wp-admin/index.php?page=my-blogs',
+			secondaryAction: i18n.translate( 'Domain Management' ),
+			secondaryActionURL: domainManagementList( selectedSite.slug )
+		} ),
+		document.getElementById( 'primary' ),
+		reduxStore
+	);
+}
+
+function isPathAllowedForDomainOnlySite( { reactContext: { pathname }, selectedSite } ) {
 	const urlPrefixesWhiteListForDomainOnlySite = [
-		domainManagementList( domainName ),
+		domainManagementList( selectedSite.slug ),
 		'/checkout/',
 	];
 
 	return urlPrefixesWhiteListForDomainOnlySite.some( path => startsWith( pathname, path ) );
 }
 
-function onSelectedSiteAvailable( context ) {
-	const selectedSite = sites.getSelectedSite();
-	const state = context.store.getState();
-
+function feedReduxStoreWithSelectedSite( { reactContext: { store: reduxStore }, selectedSite } ) {
 	// Currently, sites are only made available in Redux state by the receive
 	// here (i.e. only selected sites). If a site is already known in state,
 	// avoid receiving since we risk overriding changes made more recently.
-	if ( ! getSite( state, selectedSite.ID ) ) {
-		context.store.dispatch( receiveSite( selectedSite ) );
-	}
-
-	context.store.dispatch( setSelectedSiteId( selectedSite.ID ) );
-
-	if ( isDomainOnlySite( state, selectedSite.ID ) &&
-		! isPathAllowedForDomainOnlySite( context.pathname, selectedSite.slug ) ) {
-		page.redirect( domainManagementList( selectedSite.slug ) );
-		return false;
+	if ( ! getSite( reduxStore.getState(), selectedSite.ID ) ) {
+		reduxStore.dispatch( receiveSite( selectedSite ) );
 	}
+	reduxStore.dispatch( setSelectedSiteId( selectedSite.ID ) );
+}
 
+const RECENT_SITES_TO_KEEP = 3;
+function setRecentSitesPreferenceInReduxStore( { reactContext: { store: reduxStore }, selectedSite } ) {
 	// Update recent sites preference
+	const state = reduxStore.getState();
 	if ( hasReceivedRemotePreferences( state ) ) {
 		const recentSites = getPreference( state, 'recentSites' );
 		if ( selectedSite.ID !== recentSites[ 0 ] ) {
-			context.store.dispatch( savePreference( 'recentSites', uniq( [
+			reduxStore.dispatch( savePreference( 'recentSites', uniq( [
 				selectedSite.ID,
 				...recentSites
-			] ).slice( 0, 3 ) ) );
+			] ).slice( 0, RECENT_SITES_TO_KEEP ) ) );
 		}
 	}
+}
 
-	return true;
+function isSelectedSiteDomainOnlySite( { reactContext: { store: reduxStore }, selectedSite } ) {
```

Comment:
Can we give the state directly here. Once we do that we can make it to a selector function (and move it to the appropriate file).